src/Bicep.Core/Emit/ExpressionConverter.cs

```diff
@@ -215,9 +205,18 @@ private LanguageExpression ConvertPropertyAccess(PropertyAccessSyntax propertyAc
                         return this
                             .CreateConverterForIndexReplacement(GetResourceNameSyntax(resourceSymbol), indexExpression, propertyAccess)
                             .GetReferenceExpression(resourceSymbol, typeReference, false);
-                }
+                    default:
+                        if (LanguageConstants.OptionalDeployTimeConstantPropertyNames.Contains(propertyName) &&
+                            resourceSymbol.DeclaringResource.TryGetBody()?.SafeGetPropertyByName(propertyName) is not null)
```

Comment:
I think you should be able to simplify to:
```csharp
if (LanguageConstants.OptionalDeployTimeConstantPropertyNames.Contains(propertyName) &&
    resourceSymbol.SafeGetBodyProperty(propertyName) is {} property)
{
    // Only subsitute optional deploy-time constant properties if they are declared in the resource body.
    return this
        .CreateConverterForIndexReplacement(GetResourceNameSyntax(resourceSymbol), indexExpression, propertyAccess)
        .ConvertExpression(property.Value);
}
```

Reply:
Nice! Didn't notice that there's the `SafeGetBodyProperty` helper method.